network/server/Server.cpp

```diff
@@ -160,46 +163,67 @@ std::vector<int> Network::Server::get_connected_clients() const
 
 //-------------------------------------TCP methods------------------------------------------
 
-void Network::Server::receive_tcp_data(const std::shared_ptr<asio::ip::tcp::socket>& tcp_socket, int client_id)
+void Network::Server::receive_tcp_data(int client_id)
 {
-    auto buffer = std::make_shared<std::vector<char>>(1024);
-
-    tcp_socket->async_read_some(asio::buffer(*buffer),
-        [this, tcp_socket, buffer, client_id](const asio::error_code& error, const std::size_t bytes_transferred) {
-            if (!error && bytes_transferred > 0) {
-                buffer->resize(bytes_transferred);
-                if(std::string const message(buffer->begin(), buffer->begin() + static_cast<std::vector<char>::difference_type>(bytes_transferred)); message == "exit") {
-                    //close the socket
-                    std::cout << "Client disconnected with id: " << client_id << '\n';
-                    _tcp_sockets[client_id]->close();
+    auto socket = _tcp_sockets[client_id];
+    std::vector<char> data(PACKET_MAX_SIZE);
+    asio::async_read(*socket, asio::buffer(data),
+        [this, client_id, socket, data](const asio::error_code& error, std::size_t rc_bytes) {
+            if (!error && rc_bytes > 0) {
+                TCPPacket packet;
+                packet.setPacket(data);
+                auto data_packet = packet.getPayloadContent();
+                auto *dto = _decoder->decode(data_packet);
+                auto *tcp_message_dto = dynamic_cast<TCPMessageDTO *>(dto);
+                auto type = tcp_message_dto->getType();
+                if (type == EXIT || type == SHUTDOWN) {
                     _clients.erase(client_id);
                     _tcp_sockets.erase(client_id);
-                    return;
                 }
-                receive_tcp_data(tcp_socket, client_id); // Continue reading data from the client
-            } else if (error) {
-                std::cerr << "Client disconnected with id: " << client_id << '\n';
-                _clients.erase(client_id);
-                _tcp_sockets.erase(client_id);
+                receive_tcp_data(client_id);
```

Comment:
Add some error handling (before you didn't make a call at receive_tcp_data if there was an error